### PR TITLE
ds_prefill: Add MoE modules for DeepSeek V3/R1 prefill

### PR DESCRIPTION
**Pull request by mbezuljTT**
### Ticket
#39618

#### Summary

- Add MoE (Mixture of Experts) modules for DeepSeek V3 prefill implementation
- Implement `TtMoe`, `TtRoutedExpert`, `TtSharedExpert`, `TtFfn`, and `TtDistributedRmsNorm` as LightweightModule classes
- Consolidate model constants into `DeepSeekV3Config` for better maintainability
- Add comprehensive PCC tests comparing TT implementation against reference PyTorch models
- Follow up PR to connect with MoE gate prefill; pending #39325
- Follow up PR to use bfp8 activations and bfp4 weights in experts 

#### New TT Modules
- `TtMoe`: Main MoE module that orchestrates dispatch → routed experts → reduce → combine flow
- `TtRoutedExpert`: Handles computation for 8 routed experts per device with EP (expert parallelism)
- `TtSharedExpert`: Implements shared expert FFN with gate projection
- `TtFfn`: Standard FFN module extending TtSharedExpert
- `TtDistributedRmsNorm`: RMS normalization distributed across mesh devices

#### Reference Implementations
- Add PyTorch reference implementations for MoE components under `reference/tt/moe/`
- Create `DeepSeekV3Config` class consolidating all model constants

#### Testing
- Add PCC tests for all new modules (`test_ttnn_moe.py`, `test_ttnn_routed_expert.py`, etc.)
- Add host-side comparison tests between TT MoE and DeepSeek reference MoE
- Reorganize torch tests into `tests/torch/` folder

#### Infrastructure
- Refactor `init_helpers.py` with consolidated EP mesh mapper/composer functions
- Update validation and visualization helpers

### Checklist

- [ ] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mbezulj/2603-modules-updated)
- [ ] [blackhole-e2e-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-e2e-tests.yaml?query=branch:mbezulj/2603-modules-updated)
- [ ] [blackhole-e2e-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-e2e-tests.yaml?query=branch:mbezulj/2603-modules-updated)

### Appendix
BH linear-8 1 link perf profile
<>

BH mesh-8x4 1 link perf profile
<>

BH mesh-8x4 2 link perf profile
<>

### CODE REVIEW

File: models/demos/deepseek_v3_d_p/tt/moe/tt_routed_expert.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+                )
+                self.down_projs.append(
+                    self._create_weight_from_torch_per_device(down_weights, name=f"expert_{local_expert_idx}_down")
+                )
+        else:
+            logger.debug("Creating random weights (replicated across devices)")
+            for i in range(experts_per_chip):
+                self.gate_projs.append(self._create_random_weight((emb_dim, hidden_dim), name=f"expert_{i}_gate"))
+                self.up_projs.append(self._create_random_weight((emb_dim, hidden_dim), name=f"expert_{i}_up"))
+                self.down_projs.append(self._create_random_weight((hidden_dim, emb_dim), name=f"expert_{i}_down"))
```
[the remark this answers is not in the document]
Comment (mbezuljTT) [reply models/demos/deepseek_v3_d_p/tt/moe/tt_routed_expert.py:147]: this is on purpose.

File: models/demos/deepseek_v3_d_p/tt/moe/validation_helpers.py [hunk 2]
```diff
@@ -192,17 +198,31 @@ def validate_combine_output(
...
                 ttnn_data = ttnn_output[dispatch_group_idx, chip_id, token_id, topk_idx]
 
-                if torch.allclose(torch_data, ttnn_data, atol=atol, rtol=rtol):
-                    matches += 1
+                if use_pcc:
+                    # Use PCC for comparison
+                    _, pcc = comp_pcc(torch_data.float(), ttnn_data.float())
+                    is_match = pcc >= pcc_threshold
+                    metric_value = pcc
+                    metric_name = "PCC"
```
[the remark this answers is not in the document]
Comment (mbezuljTT) [reply models/demos/deepseek_v3_d_p/tt/moe/validation_helpers.py]: removed as it is not used

File: models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_moe.py [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+        all_routed_weights = create_torch_expert_weights(total_experts, emb_dim, hidden_dim, seed=42)
+        # Shared expert weights in HF format - TtSharedExpert handles transpose internally
+        shared_expert_weights = create_shared_expert_weights(emb_dim, hidden_dim, seed=123)
+        # Pass all routed expert weights - TtRoutedExpert distributes them across devices
+        # (each device gets unique expert weights, not replicated)
+        ttnn_routed_weights = all_routed_weights
+        profiler.end("torch_weights_creation")
+    else:
+        # When run_pcc_check=False, pass None to use internal random weight allocation
+        all_routed_weights = None
```
[the remark this answers is not in the document]
Comment (mbezuljTT) [reply models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_moe.py:236]: Expected

File: models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_routed_expert.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+                "gate_proj": torch.randn(hidden_dim, emb_dim, dtype=torch.float32) * 0.02,
+                "up_proj": torch.randn(hidden_dim, emb_dim, dtype=torch.float32) * 0.02,
+                "down_proj": torch.randn(emb_dim, hidden_dim, dtype=torch.float32) * 0.02,
+            }
+            torch_weights_list.append(weights)
+        profiler.end("torch_expert_weights_creation")
+
+        # Create torch experts for PCC validation
+        profiler.start("torch_expert_creation")
+        torch_experts = [TorchExpert(emb_dim, torch_weights=w) for w in torch_weights_list]
```
Comment (ddjekicTT) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_ttnn_routed_expert.py]: I think you missed hidden_dim as the second positional argument here..

File: models/demos/deepseek_v3_d_p/tt/moe/validation_helpers.py [hunk 5]
```diff
@@ -563,6 +583,67 @@ def compare_buffer(
...
+        num_dispatch_groups: Number of EP ranks
+        dispatch_group_size: Number of chips in dispatch group
+        experts_per_chip: Number of experts per chip
+        pcc_threshold: PCC threshold for pass/fail (default 0.99)
+        verbose: Whether to log detailed mismatch info
+
+    Returns:
+        ValidationResult with match statistics
+    """
+    from tests.ttnn.utils_for_testing import comp_pcc
```
Comment (ddjekicTT) [inline models/demos/deepseek_v3_d_p/tt/moe/validation_helpers.py]: comp_pcc is already imported at line 13 (I believe it's the same function)

File: models/demos/deepseek_v3_d_p/tt/moe/tt_shared_expert.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                shape=(emb_dim, hidden_dim), dims=(None, -1), name="gate_proj", dtype=self.weights_dtype
+            )
+            self.up_proj = self._create_random_sharded_weight(
+                shape=(emb_dim, hidden_dim), dims=(None, -1), name="up_proj", dtype=self.weights_dtype
+            )
+            self.down_proj = self._create_random_sharded_weight(
+                shape=(hidden_dim, emb_dim), dims=(None, -2), name="down_proj", dtype=self.weights_dtype
+            )
+
+    def _to_sharded_ttnn(self, torch_weight: torch.Tensor, dims: tuple, name: str, dtype: ttnn.bfloat16) -> ttnn.Tensor:
```
Comment (ddjekicTT) [inline models/demos/deepseek_v3_d_p/tt/moe/tt_shared_expert.py]: ttnn.bfloat16 is not a type.. this should probably be ttnn.DataType

File: models/demos/deepseek_v3_d_p/tt/moe/tt_shared_expert.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+            layout=ttnn.TILE_LAYOUT,
+            device=self.mesh_device,
+            dtype=dtype,
+        )
+
+        logger.debug(f"Created {name}: {tt_weight.shape}")
+        return tt_weight
+
+    def _create_sharded_weight_from_torch(
+        self, torch_weight: torch.Tensor, dims: tuple, name: str, dtype: ttnn.bfloat16
```
Comment (ddjekicTT) [inline models/demos/deepseek_v3_d_p/tt/moe/tt_shared_expert.py]: same here

File: models/demos/deepseek_v3_d_p/tt/moe/tt_shared_expert.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        """
+        Convert HuggingFace torch weight to sharded ttnn tensor.
+
+        HF/PyTorch nn.Linear weights are [out_features, in_features], but TTNN matmul(x, W)
+        expects [in_features, out_features], so we transpose before sharding.
+        """
+        torch_weight = torch_weight.T.contiguous()
+        return self._to_sharded_ttnn(torch_weight, dims, name, dtype)
+
+    def _create_random_sharded_weight(self, shape: tuple, dims: tuple, name: str, dtype: ttnn.bfloat16) -> ttnn.Tensor:
```
Comment (ddjekicTT) [inline models/demos/deepseek_v3_d_p/tt/moe/tt_shared_expert.py]: and here :D

**Review by ddjekicTT — APPROVED** [review]

**Review by dpopovTT — APPROVED** [review]
